### ISSUE TITLE
Extract variable only extracts first occurence

### ISSUE THREAD

**Issue by blayz3r** [pappasam/jedi-language-server#49]
![image](https://user-images.githubusercontent.com/12205650/99215671-a1a38c00-27a1-11eb-915e-82696885f211.png)
![image](https://user-images.githubusercontent.com/12205650/99215690-b1bb6b80-27a1-11eb-8eed-7341b3d39179.png)

**Comment by pappasam** [comment]
Thanks for submitting this issue! Would you be able to tell me:

1. Which LSP client you're using (coc, ALE, nvim-lsp, etc)?
2. Whether you ran the "extract variable" command from normal mode or from visual mode?
3. What result you would have expected?

When I run "extract variable" from where your cursor is located in the first image in normal mode, I get the following result (which seems correct) coc.nvim:

```python
p = 1
x = 1
q = x + 1

y = p + q
var_jeiiqbap = y + 1
z = x + 1 + var_jeiiqbap
```

**Comment by blayz3r** [comment]
1. ALE
2. Visually select "x+1"
3. Cursor is on the visual selection
4. Expected:

```python
p = 1
x = 1
q = var_jeiiqbap
y = p + q
var_jeiiqbap = x + 1
z = var_jeiiqbap + y +1
```

So for x + 1 wouldn't you expect all occurrences of "x+1" to be replaced?

As an aside is it possible to have an option to choose the variable name

**Comment by pappasam** [comment]
Great questions!

For "choosing the variable name", there is a known limitation for code actions within the language server protocol. See here for discussion: <https://github.com/microsoft/language-server-protocol/issues/764>. `tsserver` currently has a workaround, but I don't know how to implement this and believe it's a special case for that codebase. That said, if you have any implementation ideas, I'd be happy to discuss further in a new issue!

For "all occurences of `x + 1`" being replaced, I'm not sure that's a good idea. The most robust code actions I've seen (tsserver) currently does not support that feature. Example:

```javascript
const x = 1
const y = 2
const z = x + 1
const q = x + 1
```

Becomes:

```javascript
const x = 1
const y = 2
const newLocal = x + 1
const z = newLocal
const q = x + 1
```
Based on my understanding, it's incredibly difficult, in general, to identify sub-components of an expression across scopes throughout a code-base. It's much simpler to say "whatever is selected becomes the code associated with a new variable. If you feel strongly that this feature should exist, you might consider opening an issue with the `tsserver` maintainers and with `jedi` itself. @davidhalter might have some opinions about why this particular request may be difficult to implement for the [Script.extract_variable action](https://jedi.readthedocs.io/en/latest/docs/api.html#jedi.Script.extract_variable). Alternatively, he's demonstrated a strong connection with the force, so maybe he views this problem as easier to solve than I do :p

**Comment by blayz3r** [comment]
Thanks for the quick update, my query was mainly because Inline variable

```python
p = 1
x = 6
q = x + 1

y = p + q
z = x + 1 + y + 1
```

Does this:

```python
p = 1

q = (6) + 1

y = p + q
z = (6) + 1 + y + 1
```
This feels like an inconsistency no?

**Comment by pappasam** [comment]
tl;dr I don't think it's an inconsistency. The reason for this is somewhat deep.

You're getting pretty deep here! Based on my understanding, this is not an inconsistency. Instead, it's a fundamental property of "variables" versus "expressions". 

A "variable" takes an expression and identifies its result **uniquely** within some scope. The unique identifier is, therefore, easy to find because it's fundamentally unique, singular, and comprised of [consistent characters](https://realpython.com/python-variables/#variable-names).

An "expression" has no such unique identifier; in fact, in Python, some argue that [any value is an expression](https://stackoverflow.com/questions/4782590/what-is-an-expression-in-python#4782649)!

Put another way, this issue is requesting that we come up with some unique way of identifying an expression across a codebase. This is simple to request, but is extremely hard to implement. For example, is `x + 1` the same as `x +  1`? How about extracting it from `x + y + 1`? How about `1 + x`? For the [__add__](http://www.marinamele.com/2014/04/modifying-add-method-of-python-class.html) method, order may not particularly matter for integers, but it might matter for lists.

You might propose that we find/replace only text that perfectly matches `x + 1` and call it a day, but at best, this would only address 1 special case of your request. Instead of going down that dark, dangerous road, "extract variable" does that only thing it can reasonably do with a selected expression. It extracts only **that** expression, assigns it to a variable name, and replaces the selected expression with the variable name. It does NOT try to determine whether the selected expression is somehow logically equivalent to other expressions in the codebase.

**Comment by davidhalter** [comment]
While it would be possible to solve this exact issue, it's something I personally wouldn't want as a user. It's just too much magic. What do you do about places where you have something like `x + 2 + y`? What would you do with `1 + x`? What with `x + (1 + y)`? 

There are so many edge cases, that it just does not make sense to invest time into it. APIs should always be very well defined, otherwise people will only complain.

**Comment by pappasam** [comment]
Perfect timing @davidhalter

**Comment by blayz3r** [comment]
I think you went into the weeds  a little bit, VS-Code replaces all occurrences of the expression by default : 

![2020-11-16_16-46-32](https://user-images.githubusercontent.com/12205650/99312350-e7a53200-282b-11eb-9566-7ea4c455ca6c.gif)

Intellij does this optionally

https://youtu.be/W_IiuORF16E?t=181

I don't think there has to be algebra or factorization, but I think it's reasonable to change the expression in the entire file. I think this would make it truly the inverse of Inline variable

my 2 cents...

**Comment by davidhalter** [comment]
@pappasam Haha. Fools seldom differ :D

**Comment by pappasam** [comment]
@blayz3r haha, fair enough. I wrote a language server, so I clearly spend a lot of time in the weeds.

At the end of the day, `jedi-language-server` is designed to expose Jedi through LSP. Unless this behavior is added to Jedi, `jedi-language-server` won't support it. Since it sounds like Jedi won't support, I'll leave this issue open so people with the same question can reference / understand the current state. Thanks for all the screencasts btw, they were quite helpful!

**Comment by blayz3r** [comment]
No worries, jedi-language-server has been a pleasure to use otherwise. You can close if it's a wont' fix kinda thing.

**Comment by davidhalter** [comment]
I would be willing to discuss this feature for Jedi if someone steps up and does the work. Would need a lot of tests of course. It could be done with an additional flag "replace_all_occurrences". However I don't like what VSCode does. I like IntelliJ's approach, because it helps you with the GUI.

**Comment by pappasam** [comment]
I'll leave this open so, if @blayz3r wants to open an issue in Jedi, you can reference this discussion. I'm also not sure how we'd implement an IntelliJ-like interactive session with LSP, but that'll be a story for another day.

**Comment by blayz3r** [comment]
Closed in favor of https://github.com/davidhalter/jedi/issues/1697
